### PR TITLE
Remove leading spaces in coderefs #2907

### PR DESCRIPTION
**Pull request by jelovirt**
Add filter to remove leading whitespace from `codeblock` elements. Implements #2907.

Signed-off-by: Jarno Elovirta <jarno@elovirta.com>

### CODE REVIEW

**Comment by jelovirt** [comment]
@robander Do you think we should make the outputclass that triggers whitespace normalization configurable via plugin extension point, or just hardcode `normalize-space` for now?

File: src/main/plugins/org.dita.base/build_preprocess2_template.xml [hunk 1]
```diff
@@ -354,6 +354,7 @@ See the accompanying LICENSE file for applicable license.
           <param name="processing-mode" value="${processing-mode}" if:set="processing-mode"/>
         </filter>
         <filter class="org.dita.dost.writer.CoderefResolver" unless:set="preprocess.coderef.skip"/>
+        <filter class="org.dita.dost.writer.NormalizeCodeblock" unless:set="preprocess.codeblock.skip"/>
```
Comment (jelovirt) [inline src/main/plugins/org.dita.base/build_preprocess2_template.xml]: Is this a good guard condition property name? Maybe `preprocess.normalize-codeblock.skip` would be better?

File: src/main/java/org/dita/dost/util/SaxCache.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        }
+    }
+
+    public static class CharactersEvent implements SaxEvent {
+        public final char[] ch;
+        public final int start;
+        public final int length;
+
+        public CharactersEvent(char ch[], int start, int length) {
+            this.ch = ch;
```
Comment (jelovirt) [inline src/main/java/org/dita/dost/util/SaxCache.java]: Maybe this should make a copy of the char array, to prevent modification and allow GC to collect the array.

File: src/main/plugins/org.dita.base/build_preprocess2_template.xml (hunk 1 above)
Comment (robander) [reply src/main/plugins/org.dita.base/build_preprocess2_template.xml]: I think `preprocess.normalize-codeblock.skip` is probably better? It feels too long -- but like many of the `skip` conditions, it will probably only be seen by / used by a small number of people, so length is not a big issue.

**Comment by robander** [comment]
Missed that question 6 days ago -- I don't think it's really necessary to make that configurable, seems like if you want to use the feature you can use that token?

Looks good, only added one comment (to answer existing question from @jelovirt ).

Only thing I wonder is if there is any need to add a test case like this, where the sub-element includes newlines; I remember, way back before DITA in our SGML days, we had some parsers that reliably broke down on this - but it may not be an issue anymore. I've usually seen this as a creative way to use add a profiled line into a codeblock (filtered out for some conditions):
```
<codeblock>   test<b>
      stuff here</b>
   and finish</codeblock>
```

File: src/main/plugins/org.dita.base/build_preprocess2_template.xml (hunk 1 above)
Comment (jelovirt) [reply src/main/plugins/org.dita.base/build_preprocess2_template.xml]: Changed

**Comment by jelovirt** [comment]
Additional test with nested linebreak added

**Comment by infotexture** [comment]
@jelovirt Found an edge case where non-whitespace characters are trimmed from the first line in the codeblock after these changes.

In the “Branch filtering” section of the [DITA features in docs](http://www.dita-ot.org/dev/reference/DITA-features-in-docs.html) topic, there's a code block with hanging indentation, where the first line is indented less than the lines that follow. The source of this code block is a coderef on [Line 55](https://github.com/dita-ot/docs/blob/develop/reference/DITA-features-in-docs.dita#L55) that pulls a range of [several lines from the publishing.ditamap](https://github.com/dita-ot/docs/blob/develop/topics/publishing.ditamap#L7-L17).

```xml
<topicref href="using-dita-command.dita" keys="build-using-dita-command" locktitle="yes">
  <topicmeta>
    <navtitle>Using the <cmdname>dita</cmdname> command</navtitle>
  </topicmeta>
  <ditavalref href="../resources/expert.ditaval">
    <ditavalmeta>
      <dvrResourcePrefix>build-</dvrResourcePrefix>
    </ditavalmeta>
  </ditavalref>
```

After the changes in this PR, the hanging indentation is removed when the leading whitespace is normalized — the first two characters are missing from the first line: `<t`.

```xml
opicref href="using-dita-command.dita" keys="build-using-dita-command" locktitle="yes">
<topicmeta>
  <navtitle>Using the <cmdname>dita</cmdname> command</navtitle>
</topicmeta>
<ditavalref href="../resources/expert.ditaval">
  <ditavalmeta>
    <dvrResourcePrefix>build-</dvrResourcePrefix>
  </ditavalmeta>
</ditavalref>
```

While this particular case is certainly unusual and could be worked around in source by referencing a different range of lines with uniform indentation, the normalization process should probably guard against this sort of thing to ensure that only whitespace characters are deleted — never actual content.

**Comment by infotexture** [comment]
Found another case today where non-whitespace content is removed from the beginning of each line.

In a codeblock that pulls in the following [properties file](https://github.com/dita-ot/docs/blob/develop/samples/properties/pdf.properties) via `<coderef>`,

```properties
output.dir = out/pdf
args.gen.task.lbl = YES
args.rellinks = nofamily
```

the first 6 characters are stripped from each line when `@outputclass="normalize-space"`:

```properties
.dir = out/pdf
en.task.lbl = YES
ellinks = nofamily
```

Of course in this case there's no leading whitespace to adjust, so the problem can be solved by simply removing the `normalize-space` keyword from the outputclass.

But since the primary use case for this option is to normalize the content of external `<coderef>` sources that may change after the `<codeblock>` markup is authored, it might be useful to be able to enable normalization in advance, and trust that it will do no harm.

**Comment by infotexture** [comment]
The remaining character loss issues described above were resolved with 8f16633.
